crates/whiskers/src/grid.rs

```diff
@@ -0,0 +1,217 @@
+//!
+use vsvg::{IntoBezPathTolerance, Point};
+
+use crate::Sketch;
+
+use kurbo::PathEl;
+
+enum GridSize {
+    CellBased([f64; 2]),
+    GridBased([f64; 2]),
+}
+
+/// Stores basic grid's cell data, like column, row and canvas position
+#[derive(Clone)]
+pub struct GridCell {
+    column: usize,
+    row: usize,
+    position: Point,
+    size: [f64; 2],
+}
+
+impl IntoBezPathTolerance for GridCell {
+    fn into_bezpath_with_tolerance(self, _tolerance: f64) -> kurbo::BezPath {
+        let [width, height] = self.size;
+        let p1: kurbo::Point = self.position.into();
+        let p2: kurbo::Point = Point::new(self.position.x() + width, self.position.y()).into();
+        let p3: kurbo::Point =
+            Point::new(self.position.x() + width, self.position.y() + height).into();
+        let p4: kurbo::Point = Point::new(self.position.x(), self.position.y() + height).into();
+
+        kurbo::BezPath::from_vec(vec![
+            PathEl::MoveTo(p1),
+            PathEl::LineTo(p2),
+            PathEl::LineTo(p3),
+            PathEl::LineTo(p4),
+            PathEl::ClosePath,
+        ])
+    }
+}
+
+/// 2-dimensional square grid module
+///
+/// Borrowed from [Programing Design Systems book by Rune Madsen](https://www.programmingdesignsystems.com/), this module
+/// helps to work with 2-dimensional grids more efficiently.
+///
+/// ```rust
+/// use whiskers::grid::{Grid};
+///
+/// Grid::from_total_size([600.0, 800.0])
+///     .columns(5)
+///     .rows(10)
+///     .position(Point::new(20.0, 100.0))
+///     .spacing([10.0, 10.0])
+///     .build(sketch, |sketch, cell| {
+///         sketch.add_path(cell);
+///     });
+/// ```
+pub struct Grid {
+    dimensions: [usize; 2],
+    size: GridSize,
+    gutter: [f64; 2],
+    x_y: Point,
+    data: Vec<GridCell>,
+}
+
+impl Grid {
+    const DEFAULT_DIMENSIONS: [usize; 2] = [4, 4];
+    const DEFAULT_GUTTER: [f64; 2] = [0.0, 0.0];
+    const DEFAULT_POSITION: [f64; 2] = [0.0, 0.0];
+
+    /// Creates grid instance based on the total dimensions
+    /// given by the user.
+    pub fn from_total_size(size: [f64; 2]) -> Self {
+        Self {
+            dimensions: Grid::DEFAULT_DIMENSIONS,
+            size: GridSize::GridBased(size),
+            gutter: Grid::DEFAULT_GUTTER,
+            x_y: Point::from(Grid::DEFAULT_POSITION),
+            data: vec![],
+        }
+    }
+
+    /// Creates grid instance based on the cell dimensions
+    /// given by the user.
+    pub fn from_cell_size(size: [f64; 2]) -> Self {
+        Self {
+            dimensions: Grid::DEFAULT_DIMENSIONS,
+            size: GridSize::CellBased(size),
+            gutter: Grid::DEFAULT_GUTTER,
+            x_y: Point::from(Grid::DEFAULT_POSITION),
+            data: vec![],
+        }
+    }
+
+    /// Overrides grid's current number of rows.
+    /// By default, grid instance will have 4 rows.
+    pub fn rows(&mut self, value: usize) -> &mut Self {
```

Comment:
honestly, I didn't know once could use `fn rows(self, value: usize) -> Self` with just `self` in return instead of creating everything from scratch there, like: `Self { /* properties go here */ }` 😅